commands/command_lock.go

```diff
@@ -0,0 +1,89 @@
+package commands
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/github/git-lfs/api"
+	"github.com/github/git-lfs/config"
+	"github.com/github/git-lfs/git"
+	"github.com/spf13/cobra"
+)
+
+var (
+	lockRemote     string
+	lockRemoteHelp = "specify which remote to use when interacting with locks"
+
+	// TODO(taylor): consider making this (and the above flag) a property of
+	// some parent-command, or another similarly less ugly way of handling
+	// this
+	setLockRemoteFor = func(c *config.Configuration) {
+		c.CurrentRemote = lockRemote
+	}
+
+	lockCmd = &cobra.Command{
+		Use: "lock",
+		Run: lockCommand,
+	}
+)
+
+func lockCommand(cmd *cobra.Command, args []string) {
+	setLockRemoteFor(config.Config)
+
+	if len(args) == 0 {
+		Print("Usage: git lfs lock <path>")
+		return
+	}
+
+	latest, err := git.CurrentRemoteRef()
+	if err != nil {
+		Error(err.Error())
+		Exit("Unable to determine lastest remote ref for branch.")
+	}
+
+	path, err := lockPath(args[0])
+	if err != nil {
+		Error(err.Error())
+	}
+
+	s, resp := API.Locks.Lock(&api.LockRequest{
+		Path:               path,
+		Committer:          api.CurrentCommitter(),
+		LatestRemoteCommit: latest.Sha,
```

Comment:
Hey, @sinbad. @technoweenie and I spent a little while today talking about what you had discussed here. We came up with a few different scenarios and were unable to find a common unifying "lock granting" implementation that a server could use.

The cases that we found are:

1. Client attempts to obtain a lock on a branch that both the client and server know about, and the client has no local changes. The lock should, in this case, be granted.

2. Client attempts to obtain a lock on a branch that both the client and server know about, but there is a non-empty diff between the `HEAD` and the latest remote ref. It was not clear whether or not the lock should be granted.

3. Client attempts to obtain a lock on a branch that the server does not know about, in which case an error should be returned from the API and the client should instruct the user to push that branch, then making it identical to the first case (see above).

As I mentioned above, we had a hard time determining what locking rules should be put in place on the server, and what the API would need to look like to support it. @technoweenie suggested that we should, as a first pass, simply have a list of locks, which are granted to clients as long as there are no other locks currently in place.

I agree with his suggestion, since no one else is using Git in this way, it would be best to see how this feature gets used in the wild during the experimentation phase, and then design a ruleset (and a corresponding API) to match.

Thoughts?